cli/src/semgrep/autofix.py

```diff
@@ -130,53 +134,122 @@ def _write_contents(path: Path, contents: str) -> None:
     path.write_text(contents)
 
 
+def matches_compare(x: RuleMatch, y: RuleMatch) -> int:
+    # If paths are not the same, just order based on that.
+    # I just need a total ordering on matches.
+    if x.path < y.path:
+        return -1
+    elif x.path > y.path:
+        return 1
+    else:
+        if x.start.offset < y.start.offset:
+            return -1
+        elif y.start.offset < x.start.offset:
+            return 1
+        else:
+            if x.end.offset < y.end.offset:
+                return -1
+            elif x.end.offset > y.end.offset:
+                return 1
+            else:
+                return 0
+
+
+def matches_overlap(x: RuleMatch, y: RuleMatch) -> bool:
+    if x.path == y.path:
+        if x.start.offset < y.start.offset:
+            return x.end.offset > y.start.offset
+        elif y.start.offset < x.start.offset:
+            return y.end.offset > x.start.offset
+        elif x.start.offset == y.start.offset:
+            return True
+
+    # If they are not from the same file, they cannot overlap.
+    return False
+
+
+def deduplicate_overlapping_matches(
+    rules_and_matches: Iterable[Tuple[Rule, OrderedRuleMatchList]]
+) -> OrderedRuleMatchList:
+
+    final_matches = []
+
+    ordered_matches = sorted(
+        (match for _, matches in rules_and_matches for match in matches),
+        key=cmp_to_key(matches_compare),
+    )
+    acc = None
+
+    for match in ordered_matches:
```

Comment:
The second answer in https://stackoverflow.com/questions/10366374/what-is-the-pythonic-equivalent-to-the-fold-function-from-functional-program makes it seem like there might be a pythonic way to do this in 3.8 with the assignment operator, but sadly I think we want to support 3.7. I'm just grasping for straws really because this code so clearly wants to be a fold

Reply:
@brendongo Do you know a pythonic way to do this?